### PR TITLE
bigquery: refactor RowIterator to support fast query path

### PR DESCRIPTION
**Pull request by shollyman**
This CL significantly revamps the rowiterator.

Previously, row iteration necessitated a reference to a Table object.
For jobs, this meant we would consult job statistics to find the
reference to the destination table.

With this change, we switch to using the jobs.getQueryResults RPC to get
rows from a job, while still allowing table reads to use the existing
tabledata.list mechanism.

This change also includes support for a caching abstraction in the row
iterator, which is added but not yet leveraged.  This will be wired up
in subsequent PRs that change the query execution path.

Additionally, this PR address some other reported issues against the library as the use of jobs.getQueryResults allows the service to make decisions about the source of results (e.g. in cases where a scripting/multistatement query returns resultsets from a child job).

Fixes: #1960, #1974

### CODE REVIEW

**Comment by shollyman** [comment]
still issues with integration tests, tracking those down

File: bigquery/iterator.go [hunk 1]
```diff
@@ -145,28 +147,72 @@ func isStructPtr(x interface{}) bool {
...
+// * We can only cache data from the start of a source.
+// * We need to cache schema, rows, and next page token to effective service
+//   a request from cache.
+// * cache references are destroyed as soon as they're interrogated.  We don't
+//   want to retain the data unnecessarily, and we expect that the backend
+//   can always provide them if needed.
+type rowSource struct {
+	// TODO: figure out if job is too heavyweight, since it can carry a lot of state.
+	// What we need is a reference to the client, and a jobreference.
+	// However, we control construction so it shouldn't have query stats, etc.
```
Comment (hongalex) [inline bigquery/iterator.go:179]: Can we create a Github issue to track this?

File: bigquery/iterator_test.go [hunk 2]
```diff
@@ -35,14 +37,125 @@ type pageFetcherStub struct {
...
+			wantErr:      errNoCacheData,
+		},
+	}
+	for _, tc := range testCases {
+		gotResp, gotErr := fetchCachedPage(context.Background(), tc.inSource, tc.inSchema, tc.inStartIndex, tc.inPageSize, tc.inPageToken)
+		if gotErr != tc.wantErr {
+			t.Errorf("err mismatch.  got %v, want %v", gotErr, tc.wantErr)
+
+		} else {
+
```
Comment (hongalex) [inline bigquery/iterator_test.go:149]: nit: extra whitespace?

File: bigquery/iterator.go (hunk 1 above)
Comment (shollyman) [reply bigquery/iterator.go]: Removed the TODO.  Now that I've gotten far enough ahead in the other CL leveraging this it's not a concern.

File: bigquery/iterator_test.go (hunk 2 above)
Comment (shollyman) [reply bigquery/iterator_test.go]: Done.

File: bigquery/iterator.go [hunk 3]
```diff
@@ -145,28 +147,69 @@ func isStructPtr(x interface{}) bool {
...
+// then dispatches to either the appropriate job or table-based backend mechanism
+// as needed.
+func fetchPage(ctx context.Context, src *rowSource, schema Schema, startIndex uint64, pageSize int64, pageToken string) (*fetchPageResult, error) {
+	result, err := fetchCachedPage(ctx, src, schema, startIndex, pageSize, pageToken)
+	if err != nil {
+		if err != errNoCacheData {
+			// This likely means something more severe, like a problem with schema.
+			return nil, err
+		}
+		// If we failed to fet data from cache, invoke the appropriate service method.
```
Comment (stephaniewang526) [inline bigquery/iterator.go:203]: "fetch* data

Comment (shollyman) [reply bigquery/iterator.go:203]: addressed in the followup https://github.com/googleapis/google-cloud-go/pull/2560, this PR was submitted a while ago

**Pull request merged**